### PR TITLE
Add a new flag to allow suppressing the transitive flow of FrameworkReference at restore time. 

### LINKED ISSUE #7988: FrameworkReference suppress dependency flow via FrameworkReference.PrivateAssets
Follow up to https://github.com/dotnet/sdk/pull/3102. 

Turns out that there is a netstandard FrameworkReference which needs to be suppressed from transitive flow. 

Need to rethink https://github.com/NuGet/Home/wiki/%5BSpec%5D-FrameworkReference-in-NuGet#what-happens-in-the-future-ifwhen-frameworkreference-are-added-to-other-frameworks. 

[Spec](https://github.com/NuGet/Client.Engineering/pull/95)

//cc @rrelyea @nguerrera @dsplaisted

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/7988
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 
Introduce a PrivateAssets flag to control the dependency flow of FrameworkReference through restore. 

Alongside that remove the Pack=false flag, as PrivateAssets accomplishes the same purpose.

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  
Validation:

### CODE REVIEW

File: src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlagsUtils.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        /// Convert framework dependency flags to a friendly string.
+        /// </summary>
+        public static string GetFlagString(FrameworkDependencyFlags flags)
+        {
+            switch (flags)
+            {
+                case FrameworkDependencyFlags.All:
+                    return "all";
+                case FrameworkDependencyFlags.None:
+                    return "none";
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlagsUtils.cs]: You can fall through here.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs [hunk 2]
```diff
@@ -395,6 +409,19 @@ private static VersionRange GetVersionRange(IVsReferenceItem item)
             return VersionRange.All;
         }
 
+        /// <summary>
+        /// Get the frameworkdependencyflag based on the name.
+        /// </summary>
+        /// <param name="item"></param>
+        /// <param name="name"></param>
+        /// <returns></returns>
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs]: Doc comments are half-done.  I'd just remove them since this is a private method.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs [hunk 3]
```diff
@@ -361,6 +358,23 @@ private static DownloadDependency ToPackageDownloadDependency(IVsReferenceItem i
             return downloadDependency;
         }
 
+        private static void PopulateFrameworkDependencies(TargetFrameworkInformation tfi, IVsTargetFrameworkInfo2 targetFrameworkInfo2)
+        {
+            foreach ( var item in targetFrameworkInfo2.FrameworkReferences.Cast<IVsReferenceItem>())
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs]: Format document.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/LockFileUtils.cs [hunk 4]
```diff
@@ -591,7 +591,10 @@ private static void AddDependencies(IEnumerable<LibraryDependency> dependencies,
                 KnownLibraryProperties.FrameworkReferences,
                 out frameworkReferencesObject))
             {
-                projectLib.FrameworkReferences.AddRange((IEnumerable<string>)frameworkReferencesObject);
+                projectLib.FrameworkReferences.AddRange(
+                    ((ISet<FrameworkDependency>)frameworkReferencesObject)
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/LockFileUtils.cs:595]: Is this explicit cast guaranteed to be safe?  Should we perform an `is` check first?

File: src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlags.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGet.LibraryModel
+{
+    public enum FrameworkDependencyFlags : ushort
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlags.cs:9]: Add the `[Flags]` attribute.

Also, can this be `internal`?

File: src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlagsUtils.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        /// <summary>
+        /// Convert set of flag strings into a FrameworkDependencyFlags.
+        /// </summary>
+        public static FrameworkDependencyFlags GetFlags(IEnumerable<string> flags)
+        {
+            var result = FrameworkDependencyFlags.None;
+
+            if (flags != null)
+            {
+                foreach (var flag in flags)
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlagsUtils.cs]: There is a more maintainable implementation that uses `Enum.TryParse<T>(…)`:
```
FrameworkDependencyFlags flags = FrameworkDependencyFlags.None;

foreach (var value in values)
{
    if (Enum.TryParse<FrameworkDependencyFlags>(value, out var flag))
    {
        flags |= flag;
    }
}
```
When a new `FrameworkDependencyFlags` member is added, this method will not need to be updated too.

(Note that I also renamed `flags` to `values` since `flags` is not guaranteed to be an enumerable of valid flags.  `values` could contain `blah`.)

File: src/NuGet.Core/NuGet.LibraryModel/FrameworkDependency.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        public FrameworkDependencyFlags PrivateAssets { get; }
+
+        public FrameworkDependency(
+            string name,
+            FrameworkDependencyFlags privateAssets
+            )
+        {
+            Name = name ?? throw new ArgumentNullException(nameof(name));
+            PrivateAssets = privateAssets;
+
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.LibraryModel/FrameworkDependency.cs]: Remove blank line.

File: src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlags.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+    {
+        None = 0,
+        All = 1 << 0,
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlags.cs]: Set `All` to `ushort.MaxValue`.

File: src/NuGet.Core/NuGet.Packaging/Core/FrameworkReference.cs [hunk 9]
```diff
@@ -4,13 +4,12 @@
 using System;
 using System.Collections.Generic;
 using NuGet.Shared;
+using static NuGet.Common.ComparisonUtility;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Core/FrameworkReference.cs]: I think [this](https://github.com/NuGet/NuGet.Client/pull/2829/files#diff-1bf3c9468063aed0d10fb059f72efc43R172) is easier to understand than `using static`; you can't easily know where `FrameworkReferenceNameComparer` comes from without an IDE or searching.

File: src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlagsUtils.cs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NuGet.LibraryModel
+{
+    public class FrameworkDependencyFlagsUtils
+    {
+        public static FrameworkDependencyFlags Default = FrameworkDependencyFlags.None;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlagsUtils.cs]: `readonly`.  Does it need to be `public` or will `internal` suffice?  The same question also applies to the class.

File: src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs [hunk 11]
```diff
@@ -197,7 +197,9 @@ private static FileStream AcquireFileStream(string lockPath)
                 options: options);
         }
 
+#pragma warning disable IDE1006 // Naming Styles
         private static string _basePath;
+#pragma warning restore IDE1006 // Naming Styles
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs]: I would undo this change.  I don't see a warning/error.  If there is indeed a warning/error, let's figure it out.  If the problem is specific to me, then I'd want to think about other options (e.g.:  modifying or disabling the rule) instead of disabling it selectively.

File: src/NuGet.Core/NuGet.Common/ComparisonUtility.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace NuGet.Common
+{
+    public static class ComparisonUtility
+    {
+        public static StringComparer FrameworkReferenceNameComparer = StringComparer.OrdinalIgnoreCase;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Common/ComparisonUtility.cs]: As-is, this is a writable field.  It should be read-only.

You have two options:
1. convert into a read-only property
2. convert into a read-only field

I prefer option 2 as it's simpler.

File: src/NuGet.Core/NuGet.LibraryModel/FrameworkDependency.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    public class FrameworkDependency : IEquatable<FrameworkDependency>, IComparable<FrameworkDependency>
+    {
+        public string Name { get; }
+
+        public FrameworkDependencyFlags PrivateAssets { get; }
+
+        public FrameworkDependency(
+            string name,
+            FrameworkDependencyFlags privateAssets
+            )
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.LibraryModel/FrameworkDependency.cs]: Move `)` to end of previous line.

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs [hunk 14]
```diff
@@ -806,16 +807,32 @@ private static bool BuildTargetFrameworkNode(PackageSpec packageSpec, KeyValuePa
...
+                    var privateAssets = FrameworkDependencyFlagsUtils.Default;
+
+                    if (frameworkReference.Value is JObject frameworkReferenceToken)
+                    {
+                        if (TryGetStringEnumerable(frameworkReferenceToken["privateAssets"], out var strings))
+                        {
+                            privateAssets = FrameworkDependencyFlagsUtils.GetFlags(strings);
+                        }
+                    }
+                    frameworkReferences.Add(new FrameworkDependency(name: frameworkReference.Key, privateAssets: privateAssets));
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs]: I don't think you need the 2nd named parameter.

File: src/NuGet.Core/NuGet.LibraryModel/FrameworkDependency.cs [hunk 15]
```diff
@@ -0,0 +1,58 @@
+// Copyright(c) .NET Foundation.All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using NuGet.Common;
+
+namespace NuGet.LibraryModel
+{
+    public class FrameworkDependency : IEquatable<FrameworkDependency>, IComparable<FrameworkDependency>
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.LibraryModel/FrameworkDependency.cs]: `sealed`?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs [hunk 16]
```diff
@@ -678,20 +677,26 @@ private static void AddFrameworkReferences(PackageSpec spec, IEnumerable<IMSBuil
...
 
-            if (!frameworkInfo.FrameworkReferences.Contains(frameworkReference))
+            if (!frameworkInfo
+                .FrameworkReferences
+                .Select(f => f.Name)
+                .Contains(frameworkReference, ComparisonUtility.FrameworkReferenceNameComparer))
             {
-                frameworkInfo.FrameworkReferences.Add(frameworkReference);
+                var privateAssets = FrameworkDependencyFlagsUtils.GetFlags(MSBuildStringUtility.Split(privateAssetsValue));
+                frameworkInfo.FrameworkReferences.Add(new FrameworkDependency(name: frameworkReference, privateAssets: privateAssets));
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs]: The 2nd named parameter is unnecessary.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs [hunk 17]
```diff
@@ -347,21 +347,23 @@ private void PopulateFrameworkAssemblyReferences(PackageBuilder builder, IPackTa
...
-                if (tfmSpecificRefs.ContainsKey(targetFramework))
-                {
-                    tfmSpecificRefs[targetFramework].Add(tfmRef.Identity);
-                }
-                else
+                var frameworkShortFolderName = framework.FrameworkName.GetShortFolderName();
+                foreach (var frameworkRef in framework.FrameworkReferences.Where(e => e.PrivateAssets != FrameworkDependencyFlags.All))
                 {
-                    tfmSpecificRefs.Add(targetFramework, new HashSet<string>(FrameworkReference.FrameworkReferenceNameComparer) { tfmRef.Identity });
+                    if (tfmSpecificRefs.ContainsKey(frameworkShortFolderName))
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs]: There's a more efficient way:
```
if (tfmSpecificRefs.TryGetValue(frameworkShortFolderName, out ISet<string> frameworkRefNames)
{
    frameworkRefNames.Add(frameworkRef.Name);
}
else
{
    ...
}
```

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs [hunk 18]
```diff
@@ -806,16 +807,32 @@ private static bool BuildTargetFrameworkNode(PackageSpec packageSpec, KeyValuePa
...
             {
-                IEnumerable<string> frameworkRefArray;
-                if (TryGetStringEnumerableFromJArray(frameworkRefProperty, out frameworkRefArray))
+                foreach (var frameworkReference in frameworkRefs)
                 {
-                    frameworkReferences.AddRange(frameworkRefArray);
+                    if (string.IsNullOrEmpty(frameworkReference.Key))
+                    {
+                        throw FileFormatException.Create(
+                                   "Unable to resolve frameworkReference ''.",
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs]: Shouldn't this be a localizable resource string?  Also, it looks like there should be a replacement token in there, like `'{0}'`.

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs [hunk 19]
```diff
@@ -806,16 +807,32 @@ private static bool BuildTargetFrameworkNode(PackageSpec packageSpec, KeyValuePa
...
+                                   frameworkReference.Value,
+                                   packageSpecPath);
+                    }
+
+                    var privateAssets = FrameworkDependencyFlagsUtils.Default;
+
+                    if (frameworkReference.Value is JObject frameworkReferenceToken)
+                    {
+                        if (TryGetStringEnumerable(frameworkReferenceToken["privateAssets"], out var strings))
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:830]: Why is this `privateAssets` but [this](https://github.com/NuGet/NuGet.Client/pull/2829/files#diff-bb46f9dc27153194714455bf3f388f57R64) is `PrivateAssets`?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/LockFileUtils.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/LockFileUtils.cs:595]: It is guaranteed to be safe. It's nuget internal implementation detail. This is how we pass around project specific information in project-to-project restore.

File: src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlags.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.LibraryModel/FrameworkDependencyFlags.cs:9]: nope, it's used across different assemblies. Used in both pack & restore.

File: src/NuGet.Core/NuGet.Packaging/Core/FrameworkReference.cs (hunk 9 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/Core/FrameworkReference.cs]: Not sure what you are proposing. 

```cs
using NuGet.Common; 

//
    return ComparisonUtility.FrameworkReferenceNameComparer.Equals(Name, other.Name);
//
```

?

File: src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs (hunk 11 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs]: Fixed on my end. 

The editor config seems right.

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs (hunk 18 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs]: It's an internal error that improves debugability. 

The FileFormatException knows how to append the value (2nd param) and the spec path (3rd param).

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs (hunk 19 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:830]: It's a pattern in our package spec.
Inner objects are camelcased.

**Comment by nkolev92** [comment]
🔔 Ready for another look. 
@dtivel

File: src/NuGet.Core/NuGet.Packaging/Core/FrameworkReference.cs (hunk 9 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Packaging/Core/FrameworkReference.cs]: Yes, exactly!

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs (hunk 18 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs]: Remove:  ` ''`.

**Review by dtivel — APPROVED** [review]
